### ISSUE TITLE
MacType profile not applying on reboot

### ISSUE THREAD

**Issue by sammilucia** [snowie2000/mactype#441]
In years of using MacType I haven't seen this issue, not sure where to start looking...

### To reproduce:
1. Use Windows 8.1 x64
1. Use MacType in Service Mode as usual
1. Restart Windows
1. Log in
1. MacType is not applied to any software
1. Run MacWiz.exe
1. Re-select Service Mode, and the same profile
1. Close and re-open software
1. MacType is now applied

### Expected outcome:
- Mactype is applied automatically after restarting Windows

### Observed outcome:
- MacType isn't applied until MacWiz.exe is re-run

**Comment by sammilucia** [comment]
mactype.ini
![image](https://user-images.githubusercontent.com/3295286/46253262-fac2df80-c4b5-11e8-828b-f6cef04303ee.png)

**Comment by snowie2000** [comment]
There is a deadlock bug in the new tray app and it is much more likely to trigger in service mode.
I'll upload a fixed file later today.

**Comment by snowie2000** [comment]
[MacTray.zip](https://github.com/snowie2000/mactype/files/2431656/MacTray.zip)
Okay, here is the promised file. Hope it can fix the problem.

**Comment by sammilucia** [comment]
Unfortunately it's still occurring for me in service mode 

Straight after logging in:
![2018-10-01 12_16_37-pushbullet - your devices working better together - cent browser](https://user-images.githubusercontent.com/3295286/46267612-104b0e80-c575-11e8-8c18-49d647efa0cf.jpg)

After re-running MacWiz.exe next > next >finish (no settings changed
![image](https://user-images.githubusercontent.com/3295286/46267603-06291000-c575-11e8-8845-75461ec348bc.png)

**Comment by snowie2000** [comment]
@sammilucia Okay, so it only happens in service mode?
Could you please do a memory dump of the mactray.exe when the rendering stopped so that I can figure out what's going on inside it?

**Comment by sammilucia** [comment]
ServiceMode - I'm not sure, I'll disable secure boot and find out
Memdump - sure thing

**Comment by sammilucia** [comment]
@snowie2000 yes it still does it in registry mode... No change. Most bizarre!

**Comment by snowie2000** [comment]
Okay, that's cool.
Then there is, in theory, nothing to do with the mactray app.  It should be the problem with mactype itself.
Here is one thing you can try:
Download the process explorer from Microsoft.
Open it and select explorer.exe or any other application that you expected it to be rendered and search through its loaded dll list for MacType related Dlls. 
![qq 20181009210212](https://user-images.githubusercontent.com/11767189/46671322-0c6b5c00-cc07-11e8-836a-5a895e95c364.png)
You're expecting to see at least mactype.dll (mactype64.dll) and easyhk32(64).dll. If not, either the registry mode doesn't work at all or you enabled include-only mode.

**Comment by sammilucia** [comment]
@snowie2000 done in Registry Mode straight after a restart. Chrome is currently not rendering with MacType.
![image](https://user-images.githubusercontent.com/3295286/46708776-eafe8480-cc87-11e8-81b8-8c963c99734e.png)

MacWiz process manager seems to think it's running?
![image](https://user-images.githubusercontent.com/3295286/46708806-1bdeb980-cc88-11e8-8f5b-fc53908593ba.png)

**Comment by snowie2000** [comment]
@sammilucia Only chrome.exe is not rendered or everything is not rendered?
What is the parent process of chrome?

**Comment by sammilucia** [comment]
good question... hmm...

## renders correctly
explorer.exe -> procexp64.exe
services.exe -> svchost.exe -> mullvad
explorer.exe -> fireworks.exe
explorer.exe -> explorer windows / window titles
explorer.exe -> nexusclient.exe
explorer.exe -> greenshot.exe
services.exe -> tinywall.exe
explorer.exe -> phraseexpress.exe
explorer.exe -> flashbackplayer.exe
explorer.exe -> startisbackcfg.exe
explorer.exe -> flux.exe
explorer.exe -> mpc-hc64.exe
explorer.exe -> macwiz.exe

### doesn't render correctly
explorer.exe -> mactuner.exe (mactype.core.dll, mactype.dll, easyhk32.dll)
explorer.exe -> chrome.exe -> chrome.exe (mactype64.core.dll, mactype64.dll, easyhk64.dll)
explorer.exe -> brackets.exe -> brackets.exe (UI=yes, content=no... mactype.core.dll, mactype.dll, easyhk32.dll)
explorer.exe -> githubdesktop.exe -> githubdesktop.exe (mactype64.core.dll, mactype64.dll, easyhk64.dll)
explorer.exe -> loot.exe -> loot.exe (mactype.core.dll, mactype.dll, easyhk32.dll)

it looks like all the processes (except mactuner.exe) that aren't rendering correctly, are the ones that are spawning their own child processes:
![image](https://user-images.githubusercontent.com/3295286/46722831-05535500-ccbe-11e8-8083-78e08ab37018.png)

brackets UI is rendered correctly but content is not (assuming content is in the brackets.exe child process):
![2018-10-10 18_36_05-manifest json trellox - brackets](https://user-images.githubusercontent.com/3295286/46721859-a7be0900-ccbb-11e8-93cf-9e5647a5a31b.jpg)

**Comment by sammilucia** [comment]
chrome.exe is centbrowser of course btw @snowie2000

**Comment by snowie2000** [comment]
mactunner is naturally unrenderable. It disables mactype from the internal control center.

**Comment by sammilucia** [comment]
yes thought so. wonder what's going on with the other 3 apps and child processes then..?

**Comment by sammilucia** [comment]
Nope! Different profiles (Default, iOS) do the same thing. In addition my profile (Clean) hasn't changed, the only thing that changed was I reinstalled my Win8.1 x64.

**Comment by sammilucia** [comment]
duplicate #320?

**Comment by snowie2000** [comment]
@sammilucia I don't know, seems not related. That's a problem with the autostart of the tray app.

**Comment by sammilucia** [comment]
seems like #320 is probably not a bug but a windows / UAC / AV problem, i'll keep it closed, but feel free to reopen

**Comment by sammilucia** [comment]
@snowie2000 I have new info about MacType not auto-applying on my Win 8.1 x64 in Registry Mode or Service Mode.

In either of those modes MacType starts at startup, but as per my Process Explorer screenshot, MacType doesn't load into subprocesses (like Cent Browser tabs)

However in Tray mode it does! In either Compatibility Mode or Standalone Mode, if I set MacTray to start with Windows, MacType works for all subprocesses straight away.

In Registry and Service Mode after a restart, it doesn't load into sub processes, until I've run MacType Wizard (and simply click Next > Finish without changing anything)

I realise this is really weird, but does that provide any clues?

**Comment by snowie2000** [comment]
I have no clue why it happened. Try the new version to see if it persists.

**Comment by snowie2000** [comment]
I got an idea for that. Wait for the version I'll release tonight. (Don't use the version I submitted last night, it's buggy).

**Comment by snowie2000** [comment]
@sammilucia Could you try it for me?
After you reboot, click start menu->Run->mactype.dll，see if it gives you an error like
> Windows cannot find 'mactype.dll' Make sure you typed the name correctly... bahlabahla

**Comment by sammilucia** [comment]
Is there a new private build?

Tried with the current beta straight after reboot and got:
![image](https://user-images.githubusercontent.com/3295286/48248508-82055a80-e447-11e8-8653-44850edcb692.png)

... This was with MacType in Tray Mode (as Administrator).

**Comment by Mraru** [comment]
I (also with my classmate) have the same problem.

I use MacType in Registry Mode and my system is Windows 10 x64 Ver.1089, everytime when I restart my computer I have to run "MacTray.exe" or "MacWiz.exe" manually to apply MacType.

**Comment by txbsywcq** [comment]
Hello, same issue:
windows 8.1 x86
mactype v20170628
centbrowser v3.6.8.99 and v3.5.3.25

.ini:
DirectWrite=1
FontSubstitutes=1

Use MacType in Service Mode or registry mode
Restart Windows
Log in
Run total commander, Everything, 火绒, centbrowser
MacType is applied to any software except centbrowser and 火绒's hipsmain.exe
Find mactype.dll, mactype.core.dll, easyhk32.dll loaded in chrome.exe
Find mactype.dll, mactype.core.dll, easyhk32.dll not loaded in hipsmain.exe
Run MacWiz.exe and DO NOTHING
Close and re-open centbrowser
MacType is now applied to centbrowser
Close and re-open 火绒
MacType is not applied to 火绒's hipsmain.exe
In MacWiz's process manager, hipsmain.exe status disabled and can not start

**Comment by snowie2000** [comment]
hipsmain is an anti-malware software which definitely has self-protection builtin. It's designed to not let anything to hijack it.

This problem looks much more common than I thought.

**Comment by snowie2000** [comment]
@sammilucia @txbsywcq I would like you to try the software `PCHunter` to check for the hooks in the affected applications.
Since it is a Chinese site, I'll just give you the link. http://www.xuetr.com/download/PCHunter_free.zip

After downloading the software, launch the one matching your arch, and do as the following picture:
![untitled](https://user-images.githubusercontent.com/11767189/48537252-fc245c00-e8eb-11e8-958a-13c183a1869e.png)
The green box on the right-hand side is the hooks you're expected to see and are necessary to make MacType functional. Check all the affected processes to see if they are present.

Thank you.

**Comment by txbsywcq** [comment]
![image](https://user-images.githubusercontent.com/44917162/48675019-3b71d780-eb8e-11e8-9042-dfa455a9505d.png)
kernel32.dll vs. kernelbase.dll

**Comment by txbsywcq** [comment]
after run macwiz.exe and re-open centbrowser:

![image](https://user-images.githubusercontent.com/44917162/48675135-167e6400-eb90-11e8-8a38-6946b827f663.png)

**Comment by snowie2000** [comment]
Please check the one doesn't work but with mactype dll loaded.

**Comment by txbsywcq** [comment]
![image](https://user-images.githubusercontent.com/44917162/48675346-c05ef000-eb92-11e8-8d5a-44e069a5c7eb.png)
![image](https://user-images.githubusercontent.com/44917162/48675353-d7054700-eb92-11e8-989f-3ff6249c1388.png)
![image](https://user-images.githubusercontent.com/44917162/48675355-e71d2680-eb92-11e8-93ca-b66290890f25.png)

**Comment by txbsywcq** [comment]
The first pic: MacType is not applied to centbrowser;
After running macwiz.exe, the second pic: MacType is applied to centbrowser;

**Comment by txbsywcq** [comment]
Sorry, what do you mean about the process doesn't work but with mactype dll loaded?

**Comment by txbsywcq** [comment]
Tampermonkey process:

![image](https://user-images.githubusercontent.com/44917162/48675569-6b70a900-eb95-11e8-96cb-47d46e0d82c9.png)
![image](https://user-images.githubusercontent.com/44917162/48675584-9b1fb100-eb95-11e8-8db1-ef9e00abd728.png)
![image](https://user-images.githubusercontent.com/44917162/48675592-b094db00-eb95-11e8-855f-8ed54ec23214.png)

**Comment by sammilucia** [comment]
@snowie2000 hi sorry for the delay I'm just in the middle of launching a company (!) I'll complete this in the next couple of days.

**Comment by snowie2000** [comment]
@sammilucia It's okay as the MacType project is used to evolve slowly 🤣🤣🤣

@txbsywcq Is the picture the situation when the MacType didn't work? 'Cause it looks so normal.

**Comment by sammilucia** [comment]
@sammilucia not any more!! 😂 😜

**Comment by txbsywcq** [comment]
@snowie2000 The first and the third picture i post is the situation when mactype didn't work on centbrowser, the second picture i post is the situation when mactype work.

**Comment by snowie2000** [comment]
> @snowie2000 I have new info about MacType not auto-applying on my Win 8.1 x64 in Registry Mode or Service Mode.
> 
> In either of those modes MacType starts at startup, but as per my Process Explorer screenshot, MacType doesn't load into subprocesses (like Cent Browser tabs)
> 
> However in Tray mode it does! In either Compatibility Mode or Standalone Mode, if I set MacTray to start with Windows, MacType works for all subprocesses straight away.
> 
> In Registry and Service Mode after a restart, it doesn't load into sub processes, until I've run MacType Wizard (and simply click Next > Finish without changing anything)
> 
> I realise this is really weird, but does that provide any clues?

@txbsywcq Your result conflicts with what @sammilucia observed, as he stated that MacType *DIDN'T* load into sub-processes of the centbrowser opposed to what you witnessed.

**Comment by sammilucia** [comment]
@snowie2000 i'm a giirrrrlllll 😜

**Comment by snowie2000** [comment]
😝😝😝😝😝😝😝😝🙏🏻👱🏻‍♀️👱🏼‍♀️👱🏽‍♀️👱🏾‍♀️👱🏿‍♀️

**Comment by sammilucia** [comment]
😂😂 @snowie2000 lol

okay so this is a chrome.exe subprocess immediately after reboot in registry mode, with the DirectWrite test version of MacType:
![image](https://user-images.githubusercontent.com/3295286/48763030-40ed3000-ecfc-11e8-8628-230cd4986181.png)

and chrome.exe after running macwiz.exe:
![2018-11-20 19_33_06-yusxyzbzt](https://user-images.githubusercontent.com/3295286/48763164-8b6eac80-ecfc-11e8-9fbc-181f0aa484b0.jpg)

But here's the problem ... it's not doing it any more! MacType is properly installing after reboot!

**Comment by sammilucia** [comment]
Yep it's working perfectly in service and registry mode. That's so weird, I tested it right after I installed the DirectWrite version and it wasn't working. The only thing that's changed is I installed Adobe CS6 which installed a bunch of Visual C packages ???
![image](https://user-images.githubusercontent.com/3295286/48763432-38e1c000-ecfd-11e8-9cf8-4db0d25cdb83.png)

**Comment by snowie2000** [comment]
So the problem is gone and everything is working flawlessly now?  Even after a reboot?  Everything else is still the same excepted those runtimes you installed?

**Comment by snowie2000** [comment]
Oops...

**Comment by sammilucia** [comment]
¯\_(ツ)_/¯

**Comment by snowie2000** [comment]
Mactype is compiled dependency free, including easyhook. So installing runtimes should makes no difference.

**Comment by sammilucia** [comment]
Yeah.... really weird... It does sound like @txbsywcq has seen the same problem though

**Comment by txbsywcq** [comment]
On my win 8.1 x86 lite version, MacType is applied to any software except centbrowser after reboot.
I wonder if it's because lacking some font file, because @sammilucia said she installed Adobe CS6, i think that will install many font to system.

**Comment by snowie2000** [comment]
Could you please try chrome or cent browser with directwrite enabled?

**Comment by txbsywcq** [comment]
![2018-11-20 23-7-56](https://user-images.githubusercontent.com/44917162/48782454-31c8ab00-ed19-11e8-8676-eb79e6ded187.png)

**Comment by snowie2000** [comment]
OK and I noticed a problem. You and @sammilucia have all turned the cleartype off.  please turn it back on and try again to see if it works.

**Comment by txbsywcq** [comment]
Done, and yes it's because cleartype off.

**Comment by snowie2000** [comment]
So you mean it's working now?

**Comment by snowie2000** [comment]
Very good so we found the root of the cause.  I'll check out the details tomorrow. 
 Thank you very much

**Comment by sammilucia** [comment]
How do I turn cleartype on? (How did I turn it off?? 😜)

**Comment by txbsywcq** [comment]
search cleartype in control panel.

**Comment by snowie2000** [comment]
![qq 20181121100809](https://user-images.githubusercontent.com/11767189/48814415-96b1ee80-ed75-11e8-87b4-aabb8811c44d.png)
You can find cleartype tuner in the display property dialog.

**Comment by sammilucia** [comment]
Yes I did that's why I'm confused ... Mine says its enabled
![image](https://user-images.githubusercontent.com/3295286/48815635-139d9200-ed8f-11e8-9137-99406c480702.png)

**Comment by snowie2000** [comment]
@sammilucia That's why your Mactype is working now.
Look at the picture here https://github.com/snowie2000/mactype/issues/441#issuecomment-428483840, the ClearType was definitely off.

**Comment by sammilucia** [comment]
HMMM, let me test, I'll disable, leave MacType in Registry Mode and restart.

**Comment by snowie2000** [comment]
I tested it on Windows 10 with ClearType off, but it still works.
Maybe things are different on the latest Windows.

As far as I know, turn on/off ClearType alters the way GDI works.
Normally, all the text APIs will eventually call `ExtTextOutW` for final drawing, but if you turn ClearType off, some of the APIs will directly interact with the kernel and paint text without calling `ExtTextOutW`.
Because the way MacType works depends on this behavior, you have to keep ClearType on even it doesn't do any output.

**Comment by sammilucia** [comment]
Ahh right!

Okay I disabled ClearType, changed to Registry Mode, restarted ... The problem appeared!

I wonder if MacWiz should enable ClearType on Windows XP - 8.1 to avoid this problem in the future?

**Comment by snowie2000** [comment]
Of course I can, but some users just thought it's not necessary to have both MacType and ClearType turned on and decided to turn it back off.

**Comment by sammilucia** [comment]
Maybe it just needs a warning
